File: src/main/java/bad/robot/radiate/ui/PotentiallySingleRowOfTiles.java
package bad.robot.radiate.ui;

import static java.lang.String.format;

class PotentiallySingleRowOfTiles implements Tiles {

    private final int numberOfTiles;
    private int rows;
    private int columns;

    public static PotentiallySingleRowOfTiles tiles(int numberOfTiles) {
        return new PotentiallySingleRowOfTiles(numberOfTiles);
    }

    private PotentiallySingleRowOfTiles(int numberOfTiles) {
        this.numberOfTiles = numberOfTiles;
        for (int i = 1; i <= Math.sqrt(numberOfTiles); i++) {
            if (numberOfTiles % i == 0) {
                this.rows = i;
                this.columns = numberOfTiles / i;
            }
        }
    }

    @Override
    public int rows() {
        return rows;
    }

    @Override
    public int columns() {
        return columns;
    }

    @Override
    public String toString() {
        return format("%d tiles (%d, %d)", numberOfTiles, rows(), columns());
    }
}

File: src/main/scala/bad/robot/radiate/ui/PotentiallySingleRowOfTiles.scala
package bad.robot.radiate.ui

import scala.math._

object PotentiallySingleRowOfTiles {
  def tiles(numberOfTiles: Int) = new PotentiallySingleRowOfTiles(numberOfTiles)
}

case class PotentiallySingleRowOfTiles(numberOfTiles: Int) extends Tiles {
  private var _rows = 0
  private var _columns = 0

  for (i <- 1 to sqrt(numberOfTiles).toInt) {
    if (numberOfTiles % i == 0) {
      _rows = i
      _columns = numberOfTiles / i
    }
  }

  val rows = _rows
  val columns = _columns
  
  override def toString: String = {
    s"$numberOfTiles tiles ($rows, $columns)"
  }
}